src/imports/background/import-item-creation.js

```diff
@@ -0,0 +1,195 @@
+     * @param {(item: any) => any} transform Opt. transformformation fn turning current iterm into import item structure.
+     * @param {(url: string) => bool} alreadyExists Opt. checker function to check against existing data.
+     * @return {Map<string, any>} Filtered version of `items` array made into a Map of encoded URL strings to import items.
+     */
+    return (items, transform = f => f, alreadyExists) =>
+        items.reduce((accItems, currItem) => {
+            if (!isLoggable(currItem) || !isNotBlacklisted(currItem)) {
+                return accItems
+            }
+
+            try {
+                // Augment the item with encoded URL for DB check and item map
+                const encodedUrl = encodeUrl(currItem.url, false)
+
+                // Filter out items already existing in DB, if specified
+                if (alreadyExists == null || !alreadyExists(encodedUrl)) {
+                    accItems.set(encodedUrl, transform(currItem))
+                }
+                return accItems
+            } catch (error) {
+                // URI malformed; don't add to list
+                return accItems
+            }
+        }, new Map())
+}
+
+/**
+ * @param {number} [maxResults=9999999] The max amount of items to grab from browser.
+ * @param {number} [startTime=0] The time to start search from for browser history.
+ * @returns {Array<browser.history.HistoryItem>} All history items in browser filtered by URL.
+ */
+const getHistoryItems = (maxResults = 9999999, startTime = 0) =>
+    browser.history.search({
+        text: '',
+        maxResults,
+        startTime,
+    })
+
+/**
+ * @param {number} [maxResults=100000] The max amount of items to grab from browser.
+ * @returns {Array<browser.bookmarks.BookmarkTreeNode>} All bookmark items in browser filtered by URL.
+ */
+const getBookmarkItems = (maxResults = 100000) =>
+    browser.bookmarks.getRecent(maxResults)
+
+export async function getOldExtItems() {
+    let bookmarkUrls = new Set()
+    const {
+        [OLD_EXT_KEYS.INDEX]: index,
+        [OLD_EXT_KEYS.BOOKMARKS]: bookmarks,
+    } = await browser.storage.local.get({
+        [OLD_EXT_KEYS.INDEX]: { index: [] },
+        [OLD_EXT_KEYS.BOOKMARKS]: '[]',
+    })
+
+    if (typeof bookmarks === 'string') {
+        try {
+            bookmarkUrls = new Set(JSON.parse(bookmarks).map(bm => bm.url))
+        } catch (error) {}
+    }
+
+    const transform = transformOldExtDataToImportItem(bookmarkUrls)
+    const importItems = []
+
+    // Only attempt page data conversion if index + bookmark storage values are correct types
+    if (index && index.index instanceof Array) {
+        // NOTE: There is a bug with old ext index sometimes repeating the index keys
+        //  (eg.I indexed 400 pages, index contained 43 million)
+        const indexSet = new Set(index.index)
+
+        for (const keyChunk of chunk(200)([...indexSet])) {
+            const storageChunk = await browser.storage.local.get(keyChunk)
```

Comment:
a blocking operation inside a loop seems a bad idea, I wonder if we can change this to parallel (I would need some testing about performances), something like (I'm sorry if the syntax is incorrect, I generally optimize for readability not speed)
```
const chunk200 =chunk(200)([...indexSet]
Promise.all(
	chunk200
    //assuming browser etc is a chainable promise, otherwise the standard new Promise etc
    .map(keyChunk=>Promise.resolve(browser.storage.local.get(keyChunk))) 
)
.then([...storageChunk]=>{
    for (const [index,keyChunk] of chunk200.entries()) {
	    keyChunk.forEach((key, i) => {
	        if (storageChunk[index][key] == null) {
	            return
	        }
	        importItems.push(transform(storageChunk[index][key], i))
	    }) 
    }
}
```
On the other side, I have a suspicion this code did intentionally use await to don't be blocking, is it the case?

Reply:
Yes intentionally blocking. Common way to cap the space complexity to a constant level, tradeoff being more time. 

If we use `Promise.all` here, each `browser.storage.local.get` call would be processing in "parallel", but our space usage here would be linear to the old ext index size. We had this like this until last week, then we released and it blew up everyone's browser who had sufficient amount of old ext data.

Reply:
got it. Probably worth adding a comment in the code to explain why you are doing it like that